client/ayon_core/hosts/nuke/api/utils.py

```diff
@@ -142,3 +149,160 @@ def is_headless():
+        # code.
+        with nuke.root():
+            _submit_headless_farm(node)
+
+
+def _submit_headless_farm(node):
+    context = util.collect()
+
+    success, error_report = create_error_report(context)
+
+    if not success:
+        show_message_dialog(
+            "Collection Errors", error_report, level="critical"
+        )
+        return
+
+    # Find instance for node and workfile.
+    instance = None
+    instance_workfile = None
+    indexes_to_remove = []
+    for count, Instance in enumerate(context):
+        if Instance.data["family"] == "workfile":
+            instance_workfile = Instance
+            continue
+
+        instance_node = Instance.data["transientData"]["node"]
+        if node.name() == instance_node.name():
+            instance = Instance
+        else:
+            indexes_to_remove.append(count)
+
+    if instance is None:
+        show_message_dialog(
+            "Collection Error",
+            "Could not find the instance from the node.",
+            level="critical"
+        )
+        return
+
+    # Enable for farm publishing.
+    instance.data["farm"] = True
+    instance.data["transfer"] = False
+
+    # Clear the families as we only want the main family, ei. no review etc.
+    instance.data["families"] = []
+
+    # Use the workfile instead of published.
+    publish_attributes = instance.data["publish_attributes"]
+    publish_attributes["NukeSubmitDeadline"]["use_published_workfile"] = False
+
+    # Disable version validation.
+    instance.data.pop("latestVersion")
+    instance_workfile.data.pop("latestVersion")
+
+    # Remove all other instances.
+    indexes_to_remove.sort(reverse=True)
+    for i in indexes_to_remove:
+        if 0 <= i < len(context):
+            del context[i]
+
+    # Validate
+    util.validate(context)
+
+    success, error_report = create_error_report(context)
+
+    if not success:
+        show_message_dialog(
+            "Validation Errors", error_report, level="critical"
+        )
+        return
+
+    # Extraction.
+    util.extract(context)
+
+    success, error_report = create_error_report(context)
+
+    if not success:
+        show_message_dialog(
+            "Extraction Errors", error_report, level="critical"
+        )
+        return
+
+    # Save the workfile.
+    host = registered_host()
+    host.save_file(host.current_file())
+
+    # Copy the workfile to a timestamped copy.
+    current_datetime = datetime.now()
+    formatted_timestamp = current_datetime.strftime("%Y%m%d%H%M%S")
+    base, ext = os.path.splitext(host.current_file())
+
+    directory = os.path.join(os.path.dirname(base), "farm_submissions")
+    if not os.path.exists(directory):
+        os.makedirs(directory)
+
+    filename = "{}_{}{}".format(
+        os.path.basename(base), formatted_timestamp, ext
+    )
+    path = os.path.join(directory, filename).replace("\\", "/")
+    context.data["currentFile"] = path
```

Comment:
Do we really want to hijack the publishing logic this much? This looks like an anti-pattern?

Reply:
Got any suggestions to improve it?

Reply:
I feel like:

- Collect instances
- Find the relevant instance
- Set the relevant instance attribute value (toggle the active state, etc.)
- Continue publish with only that instance active

Which would be along the lines of:

```python
# pseudocode
create_context = CreateContext(host)
for instance in create_context.instances:
    instance.data["active"] = is_the_relevant_intsance(instance)
    instance.data["render_mode"] = "farm"  # or whatever the attribute definition and value is
try:
    create_context.publish()
except PublishFailed:
    report = create_context.get_report()
    show_message_dialog("Publish failed", report, level="critical")
```

This should essentially all be handled by 'default publishing logic' - we just want to trigger it simply for this one instance, right?

Reply:
> The client has requested to have a timestamped workfile instead of rendering from the current workfile, hence the hijacked publishing process.

I mean - that sounds _just_ as useful for any farm submission. Why not implement it in the regular publishing logic?

Even more so - wasn't there like tons of logic already that allowed farm rendering to basically render from a **published workfile** which in essence is exactly that - a 'stamped' static version for the farm? (Timestamp is then the publish datetime, etc.)

Reply:
> Client does not want a published workfile to clutter the database.

This feels quite related to [this Ynput Community forum discussion on not publishing directly, but first making a 'reviewable' first](https://community.ynput.io/t/publishing-renders-from-houdini/1359/2).

Anyway, since your logic is still not removing the workfile instance, nor removing it - isn't it still being published?

Reply:
> Anyway, since your logic is still not removing the workfile instance, nor removing it - isn't it still being published?

True, did not notice that. The publishing does not actually publish the workfile when testing, and I dont know why, but I've deactivated the instance just in case.